### PR TITLE
ARROW-6144: [C++][Gandiva] Implement random functions in Gandiva

### PR DESCRIPTION
**Pull request by pprudhvi**

### CODE REVIEW

File: cpp/src/gandiva/precompiled/extended_math_ops.cc [hunk 1]
```diff
@@ -123,4 +124,24 @@ int64 truncate_int64_int32(int64 in, int32 out_scale) {
...
+  std::uniform_real_distribution<> dis(0.0, 1.0);
+  return dis(gen);
+}
+
+FORCE_INLINE
+float64 random_int32_internal(int32 seed, bool seed_validity) {
+  if (!seed_validity) {
+    seed = 0;
+  }
+  int64 seed64 = static_cast<int64>(seed);
```
Comment (bkietz) [inline cpp/src/gandiva/precompiled/extended_math_ops.cc]: could you add a comment describing what this seed manipulation accomplishes?

**Review by bkietz — CHANGES_REQUESTED** [review]

File: cpp/src/gandiva/precompiled/extended_math_ops.cc (hunk 1 above)
Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/extended_math_ops.cc]: I wanted to scramble the seed(x) in such a way that the result is a positive number and is not the same for x or -x. I looked up how java.util.Random scrambles the seed; lifted the constants from there

File: cpp/src/gandiva/precompiled/extended_math_ops.cc [hunk 2]
```diff
@@ -15,6 +15,7 @@
 // specific language governing permissions and limitations
 // under the License.
 
+#include <random>
```
Comment (praveenbingo) [inline cpp/src/gandiva/precompiled/extended_math_ops.cc]: why do we need this here..

File: cpp/src/gandiva/random_generator_holder.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    seed64 = (seed64 ^ 0x00000005DEECE66D) & 0x0000ffffffffffff;
+    generator_.seed(static_cast<uint64_t>(seed64));
+  }
+
+  RandomGeneratorHolder() {
+    std::random_device rd;
+    generator_.seed(rd());
+  }
+
+  std::mt19937_64 generator_;
```
Comment (praveenbingo) [inline cpp/src/gandiva/random_generator_holder.h]: have we tested the perf for this? looks like this is the slowest of the three random generators
https://en.cppreference.com/w/cpp/numeric/random

do we need anything better than the fibonacci one..

File: cpp/src/gandiva/random_generator_holder.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+ private:
+  explicit RandomGeneratorHolder(int seed) {
+    int64_t seed64 = static_cast<int64_t>(seed);
+    seed64 = (seed64 ^ 0x00000005DEECE66D) & 0x0000ffffffffffff;
+    generator_.seed(static_cast<uint64_t>(seed64));
+  }
+
+  RandomGeneratorHolder() {
+    std::random_device rd;
```
Comment (praveenbingo) [inline cpp/src/gandiva/random_generator_holder.h]: how costly is it to create one per call..should it be static? is it thread safe..

File: cpp/src/gandiva/random_generator_holder.h [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    generator_.seed(rd());
+  }
+
+  std::mt19937_64 generator_;
+  std::uniform_real_distribution<> distribution;
```
Comment (praveenbingo) [inline cpp/src/gandiva/random_generator_holder.h]: can you please mention the range this bounds to.

File: cpp/src/gandiva/precompiled/extended_math_ops.cc (hunk 2 above)
Comment (pprudhvi) [reply cpp/src/gandiva/precompiled/extended_math_ops.cc]: cleaned up.

File: cpp/src/gandiva/random_generator_holder.h (hunk 3 above)
Comment (pprudhvi) [reply cpp/src/gandiva/random_generator_holder.h]: this appears to be as fast as the fibonacci one for our purposes

File: cpp/src/gandiva/random_generator_holder.h (hunk 5 above)
Comment (pprudhvi) [reply cpp/src/gandiva/random_generator_holder.h]: done. Thanks

**Comment by praveenbingo** [comment]
@pprudhvi could you please restart travis on this PR. looks like travis had some issue around the time this was pushed..

**Review by praveenbingo — COMMENTED** [review]
please also add a test from java..usually it uncovers all issues when we use the holder for doing generation.

**Review by praveenbingo — APPROVED** [review]

**Pull request closed without merging**